Add Mistral AI Chat Completion support to Inference Plugin

Jan-Kazlouski-elastic opened the pull request:
Change to existing Mistral AI  provider integration allowing completion (both streaming and non-streaming) and chat_completion (only streaming) to be executed as part of inference API.
Changes were tested against next models:

- mistral-large-latest
- mistral-small-latest

Notes:

- Mistral returns at least 5 different formats of non-streaming errors. While in their [documentation](https://docs.mistral.ai/api/#tag/chat/operation/chat_completion_v1_chat_completions_post) only 1 type is mentioned. However MistralErrorEntity handles error formats for 4 most common errors: Unauthorized, Bad Request, Not Found, Unprocessable Entity.
- Format of streaming errors is not defined in Mistral documentation and wasn't received during testing. Assuming mid stream errors follow Open AI format.
- Changes were made to common OpenAI Response Handler allowing it to handle more error codes. That might affect behavior for other providers, but it is better solution than having duplication.
- Mistral AI doesn't recognize "stream_options" field despite it being present in OpenAI schema. I added provider-based solution to not include it for Mistral.
- Task settings not being passed as parameters but not being used is already present solution that might need improvement, e.g. not passing it as parameter.

Examples of RQ/RS from local testing:
<details><summary>Create Completion Endpoint</summary>

Success:
```
RQ:
PUT {{base-url}}/_inference/completion/mistral-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "inference_id": "mistral-completion",
    "task_type": "completion",
    "service": "mistral",
    "service_settings": {
        "model": "mistral-small-latest",
        "rate_limit": {
            "requests_per_minute": 240
        }
    }
}
```

Unauthorized:
```
RQ:
PUT {{base-url}}/_inference/completion/mistral-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{invalid-mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Received an authentication error status code for request from inference entity id [mistral-completion] status [401]. Error message: [Unauthorized]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Received an authentication error status code for request from inference entity id [mistral-completion] status [401]. Error message: [Unauthorized]"
        }
    },
    "status": 400
}
```

Not Found:
```
RQ:
PUT {{base-url}}/_inference/completion/mistral-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"
        }
    },
    "status": 400
}
```

Invalid Model:
```
RQ:
PUT {{base-url}}/_inference/completion/mistral-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "wrong-model-name"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Received a bad request status code for request from inference entity id [mistral-completion] status [400]. Error message: [Invalid model: wrong-model-name]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Received a bad request status code for request from inference entity id [mistral-completion] status [400]. Error message: [Invalid model: wrong-model-name]"
        }
    },
    "status": 400
}
```
</details>
<details><summary>Perform Non-Streaming Completion</summary>

Success:
```
RQ:
POST {{base-url}}/_inference/completion/mistral-completion
{
    "input": "The sky above the port was the color of television tuned to a dead channel."
}
RS:
{
    "completion": [
        {
            "result": "The sentence you've provided is the opening line of William Gibson's seminal cyberpunk novel *Neuromancer*. This vivid and evocative description sets the tone for the dystopian, high-tech, low-life world that the novel explores. The imagery of a \"dead channel\" on a television screen suggests a sense of emptiness, static, and the absence of clear signals or information, which can be seen as a metaphor for the fragmented and often chaotic nature of the future Gibson envisions.\n\nThe use of such a striking opening line is characteristic of Gibson's style, which often blends technological and cultural references to create a rich, immersive atmosphere. *Neuromancer* is known for its influence on the cyberpunk genre and its prescient exploration of themes related to artificial intelligence, virtual reality, and the digital age."
        }
    ]
}
```

Not Found:
```
RQ:
POST {{base-url}}/_inference/completion/mistral-completion
{
    "input": "The sky above the port was the color of television tuned to a dead channel."
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"
            }
        ],
        "type": "status_exception",
        "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"
    },
    "status": 404
}
```
</details>

<details><summary>Perform Streaming Completion</summary>

Success:
```
RQ:
POST {{base-url}}/_inference/completion/mistral-completion/_stream
{
    "input": "The sky above the port was the color of television tuned to a dead channel."
}
RS:
event: message
data: {"completion":[{"delta":"The"}]}

﻿event: message
data: {"completion":[{"delta":" sentence"},{"delta":" you"}]}

﻿event: message
data: {"completion":[{"delta":"'ve"}]}

﻿event: message
data: {"completion":[{"delta":" provided"}]}

﻿event: message
data: {"completion":[{"delta":" is"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" opening"}]}

﻿event: message
data: {"completion":[{"delta":" line"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" William"}]}

﻿event: message
data: {"completion":[{"delta":" Gibson"}]}

﻿event: message
data: {"completion":[{"delta":"'s"}]}

﻿event: message
data: {"completion":[{"delta":" seminal"}]}

﻿event: message
data: {"completion":[{"delta":" cyber"}]}

﻿event: message
data: {"completion":[{"delta":"punk"}]}

﻿event: message
data: {"completion":[{"delta":" novel"}]}

﻿event: message
data: {"completion":[{"delta":" *"}]}

﻿event: message
data: {"completion":[{"delta":"Ne"}]}

﻿event: message
data: {"completion":[{"delta":"u"}]}

﻿event: message
data: {"completion":[{"delta":"rom"}]}

﻿event: message
data: {"completion":[{"delta":"ancer"}]}

﻿event: message
data: {"completion":[{"delta":"*."}]}

﻿event: message
data: {"completion":[{"delta":" This"}]}

﻿event: message
data: {"completion":[{"delta":" vivid"}]}

﻿event: message
data: {"completion":[{"delta":" and"}]}

﻿event: message
data: {"completion":[{"delta":" evoc"}]}

﻿event: message
data: {"completion":[{"delta":"ative"}]}

﻿event: message
data: {"completion":[{"delta":" description"}]}

﻿event: message
data: {"completion":[{"delta":" sets"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" tone"}]}

﻿event: message
data: {"completion":[{"delta":" for"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" dyst"}]}

﻿event: message
data: {"completion":[{"delta":"op"}]}

﻿event: message
data: {"completion":[{"delta":"ian"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" high"}]}

﻿event: message
data: {"completion":[{"delta":"-tech"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" low"}]}

﻿event: message
data: {"completion":[{"delta":"-life"}]}

﻿event: message
data: {"completion":[{"delta":" world"}]}

﻿event: message
data: {"completion":[{"delta":" that"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" novel"}]}

﻿event: message
data: {"completion":[{"delta":" explores"}]}

﻿event: message
data: {"completion":[{"delta":"."}]}

﻿event: message
data: {"completion":[{"delta":" The"}]}

﻿event: message
data: {"completion":[{"delta":" imagery"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" \""}]}

﻿event: message
data: {"completion":[{"delta":"dead"}]}

﻿event: message
data: {"completion":[{"delta":" channel"}]}

﻿event: message
data: {"completion":[{"delta":"\""}]}

﻿event: message
data: {"completion":[{"delta":" on"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" television"}]}

﻿event: message
data: {"completion":[{"delta":" screen"}]}

﻿event: message
data: {"completion":[{"delta":" suggests"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" sense"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" empt"}]}

﻿event: message
data: {"completion":[{"delta":"iness"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" static"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" and"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" absence"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" clear"}]}

﻿event: message
data: {"completion":[{"delta":" signals"}]}

﻿event: message
data: {"completion":[{"delta":" or"}]}

﻿event: message
data: {"completion":[{"delta":" information"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" which"}]}

﻿event: message
data: {"completion":[{"delta":" can"}]}

﻿event: message
data: {"completion":[{"delta":" be"}]}

﻿event: message
data: {"completion":[{"delta":" seen"}]}

﻿event: message
data: {"completion":[{"delta":" as"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" metaphor"}]}

﻿event: message
data: {"completion":[{"delta":" for"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" fragmented"}]}

﻿event: message
data: {"completion":[{"delta":" and"}]}

﻿event: message
data: {"completion":[{"delta":" often"}]}

﻿event: message
data: {"completion":[{"delta":" confusing"},{"delta":" reality"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" characters"}]}

﻿event: message
data: {"completion":[{"delta":" in"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" story"}]}

﻿event: message
data: {"completion":[{"delta":".\n\n"}]}

﻿event: message
data: {"completion":[{"delta":"Gib"}]}

﻿event: message
data: {"completion":[{"delta":"son"}]}

﻿event: message
data: {"completion":[{"delta":"'s"}]}

﻿event: message
data: {"completion":[{"delta":" use"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" such"}]}

﻿event: message
data: {"completion":[{"delta":" imagery"}]}

﻿event: message
data: {"completion":[{"delta":" is"}]}

﻿event: message
data: {"completion":[{"delta":" characteristic"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" cyber"}]}

﻿event: message
data: {"completion":[{"delta":"punk"}]}

﻿event: message
data: {"completion":[{"delta":" genre"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" which"}]}

﻿event: message
data: {"completion":[{"delta":" often"}]}

﻿event: message
data: {"completion":[{"delta":" blends"}]}

﻿event: message
data: {"completion":[{"delta":" advanced"}]}

﻿event: message
data: {"completion":[{"delta":" technology"}]}

﻿event: message
data: {"completion":[{"delta":" with"}]}

﻿event: message
data: {"completion":[{"delta":" social"}]}

﻿event: message
data: {"completion":[{"delta":" decay"}]}

﻿event: message
data: {"completion":[{"delta":" and"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" sense"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" alien"}]}

﻿event: message
data: {"completion":[{"delta":"ation"}]}

﻿event: message
data: {"completion":[{"delta":"."}]}

﻿event: message
data: {"completion":[{"delta":" The"}]}

﻿event: message
data: {"completion":[{"delta":" \""}]}

﻿event: message
data: {"completion":[{"delta":"port"}]}

﻿event: message
data: {"completion":[{"delta":"\""}]}

﻿event: message
data: {"completion":[{"delta":" mentioned"}]}

﻿event: message
data: {"completion":[{"delta":" could"}]}

﻿event: message
data: {"completion":[{"delta":" refer"}]}

﻿event: message
data: {"completion":[{"delta":" to"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" space"}]}

﻿event: message
data: {"completion":[{"delta":"port"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" se"}]}

﻿event: message
data: {"completion":[{"delta":"aport"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" or"}]}

﻿event: message
data: {"completion":[{"delta":" even"}]}

﻿event: message
data: {"completion":[{"delta":" a"}]}

﻿event: message
data: {"completion":[{"delta":" data"}]}

﻿event: message
data: {"completion":[{"delta":" port"}]}

﻿event: message
data: {"completion":[{"delta":","}]}

﻿event: message
data: {"completion":[{"delta":" adding"}]}

﻿event: message
data: {"completion":[{"delta":" to"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" ambiguity"}]}

﻿event: message
data: {"completion":[{"delta":" and"}]}

﻿event: message
data: {"completion":[{"delta":" futur"}]}

﻿event: message
data: {"completion":[{"delta":"istic"}]}

﻿event: message
data: {"completion":[{"delta":" feel"}]}

﻿event: message
data: {"completion":[{"delta":" of"}]}

﻿event: message
data: {"completion":[{"delta":" the"}]}

﻿event: message
data: {"completion":[{"delta":" setting"}]}

﻿event: message
data: {"completion":[{"delta":"."}]}

﻿event: message
data: [DONE]
```

Not Found:
```
RQ:
POST {{base-url}}/_inference/completion/mistral-completion/_stream
{
    "input": "The sky above the port was the color of television tuned to a dead channel."
}
RS:
event: error
data: {"error":{"root_cause":[{"type":"status_exception","reason":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"}],"type":"status_exception","reason":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [Not Found]"},"status":404}
```
</details>

<details><summary>Create Chat Completion Endpoint</summary>

Success:
```
RQ:
PUT {{base-url}}/_inference/chat_completion/mistral-chat-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "inference_id": "mistral-chat-completion",
    "task_type": "chat_completion",
    "service": "mistral",
    "service_settings": {
        "model": "mistral-small-latest",
        "rate_limit": {
            "requests_per_minute": 240
        }
    }
}
```

Unauthorized:
```
RQ:
PUT {{base-url}}/_inference/chat_completion/mistral-chat-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{invalid-mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Received an authentication error status code for request from inference entity id [mistral-chat-completion] status [401]. Error message: [Unauthorized]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Received an authentication error status code for request from inference entity id [mistral-chat-completion] status [401]. Error message: [Unauthorized]"
        }
    },
    "status": 400
}
```

Not Found:
```
RQ:
PUT {{base-url}}/_inference/chat_completion/mistral-chat-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "mistral-small-latest"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [Not Found]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [Not Found]"
        }
    },
    "status": 400
}
```

Invalid Model:
```
RQ:
PUT {{base-url}}/_inference/chat_completion/mistral-chat-completion
{
    "service": "mistral",
    "service_settings": {
        "api_key": "{{mistral-api-key}}",
        "model": "invalid-model-name"
    }
}
RS:
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [Invalid model: invalid-model-name]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [Invalid model: invalid-model-name]"
        }
    },
    "status": 400
}
```
</details>

<details><summary>Perform Streaming Chat Completion</summary>

Success:
```
RQ:
POST {{base-url}}/_inference/chat_completion/mistral-chat-completion/_stream
{
    "model": "mistral-small-latest",
    "messages": [
        {
            "role": "user",
            "content": "What is deep learning?"
        }
    ],
    "max_completion_tokens": 2
}
RS:
event: message
data: {"id":"c79e758e9d1a4e89866c9165701496f2","choices":[{"delta":{"content":"","role":"assistant"},"index":0}],"model":"mistral-small-latest","object":"chat.completion.chunk"}

﻿event: message
data: {"id":"c79e758e9d1a4e89866c9165701496f2","choices":[{"delta":{"content":"Deep"},"index":0}],"model":"mistral-small-latest","object":"chat.completion.chunk"}

﻿event: message
data: {"id":"c79e758e9d1a4e89866c9165701496f2","choices":[{"delta":{"content":" learning"},"finish_reason":"length","index":0}],"model":"mistral-small-latest","object":"chat.completion.chunk","usage":{"completion_tokens":2,"prompt_tokens":8,"total_tokens":10}}

﻿event: message
data: [DONE]
```

Invalid Model:
```
RQ:
POST {{base-url}}/_inference/chat_completion/mistral-chat-completion/_stream
{
    "model": "invalid-model-name",
    "messages": [
        {
            "role": "user",
            "content": "What is deep learning?"
        }
    ],
    "max_completion_tokens": 2
}
RS:
event: error
data: {"error":{"code":"bad_request","message":"Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [Invalid model: invalid-model-name]","type":"mistral_error"}}

Negative Max Tokens:
RQ:
POST {{base-url}}/_inference/chat_completion/mistral-chat-completion/_stream
{
    "model": "mistral-small-latest",
    "messages": [
        {
            "role": "user",
            "content": "What is deep learning?"
        }
    ],
    "max_completion_tokens": -1
}
RS:
event: error
data: {"error":{"code":"unprocessable_entity","message":"Received an input validation error response for request from inference entity id [mistral-chat-completion] status [422]. Error message: [Input should be greater than or equal to 0]","type":"mistral_error"}}
```

Not Found:
```
RQ:
POST {{base-url}}/_inference/chat_completion/mistral-chat-completion/_stream
{
    "model": "mistral-small-latest",
    "messages": [
        {
            "role": "user",
            "content": "What is deep learning?"
        }
    ],
    "max_completion_tokens": 2
}
RS:
event: error
data: {"error":{"code":"not_found","message":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [Not Found]","type":"mistral_error"}}
```

</details>

- [X] - Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- [X] - Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- [X] - If submitting code, have you built your formula locally prior to submission with `gradle check`?
- [X] - If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- [X] - If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- [X] - If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

Jan-Kazlouski-elastic committed:
Add Mistral AI Chat Completion support to Inference Plugin

Jan-Kazlouski-elastic committed:
Add changelog file

Jan-Kazlouski-elastic committed:
Fix tests and typos

jonathan-buttner reviewed (changes requested):
Looking good. I left a few suggestions.

Could you update the description of the PR so that the example requests are formatted?

Let's also wrap them in code blocks using the three backticks. Thanks for making them collapsable sections though!

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into feature/mistral-chat-completion-integration

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersions.java

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/unified/UnifiedChatCompletionRequestEntity.java:
```diff
@@ -162,11 +166,19 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-            builder.startObject(STREAM_OPTIONS_FIELD);
-            builder.field(INCLUDE_USAGE_FIELD, true);
-            builder.endObject();
+            fillStreamOptionsFields(builder);
```
Thank you for the heads up! I inspected the changes in the attached PR. They don't affect my changes. We're good.

Jan-Kazlouski-elastic replied:
Done.

Jan-Kazlouski-elastic committed:
Refactor Mistral chat completion integration and add tests

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/MistralChatCompletionResponseHandler.java:
```diff
@@ -0,0 +1,28 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.services.openai;
```
Good catch. Moved it to appropriate package.

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/embeddings/MistralEmbeddingsModel.java:
```diff
@@ -58,6 +58,20 @@ public MistralEmbeddingsModel(MistralEmbeddingsModel model, MistralEmbeddingsSer
         setPropertiesFromServiceSettings(serviceSettings);
     }
 
+    protected void setPropertiesFromServiceSettings(MistralEmbeddingsServiceSettings serviceSettings) {
```
Don't see a reason for it to be protected. Changed to private. Same as for completions.

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/MistralChatCompletionResponseHandler.java:
```diff
@@ -0,0 +1,28 @@
+import org.elasticsearch.xpack.inference.services.mistral.response.MistralErrorResponseEntity;
+
+/**
+ * Handles non-streaming chat completion responses for Mistral models, extending the OpenAI chat completion response handler.
```
Fixed. Thanks.

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralService.java:
```diff
@@ -79,11 +98,16 @@ protected void doInfer(
-            action.execute(inputs, timeout, listener);
-        } else {
-            listener.onFailure(createInvalidModelException(model));
+        switch (model) {
```
Yes, I forgot about JDK version. Fixed now.

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiResponseHandler.java:
```diff
@@ -86,11 +88,21 @@ protected void checkForFailureStatusCode(Request request, HttpResult result) thr
             throw new RetryException(false, buildError(AUTHENTICATION, request, result));
         } else if (statusCode >= 300 && statusCode < 400) {
             throw new RetryException(false, buildError(REDIRECTION, request, result));
+        } else if (statusCode == 422) {
```
That approach will lead to duplication because while this check will have to be present in both streaming and non-streaming handler classes. Streaming handler class has this hierarchy:

MistralUnifiedChatCompletionResponseHandler
extends
OpenAiUnifiedChatCompletionResponseHandler
extends
OpenAiChatCompletionResponseHandler
extends 
OpenAiResponseHandler

Non-streaming handler has this hierarchy:
MistralCompletionResponseHandler
extends 
OpenAiChatCompletionResponseHandler
extends 
OpenAiResponseHandler

Closest common ancestor is OpenAiChatCompletionResponseHandler, but streaming handler also needs to take in logic from OpenAiUnifiedChatCompletionResponseHandler. 
Do we really want to introduce such duplication? It will be literally 1 to 1 because errors are the same. Not very good practice. Please correct me if i'm wrong.

>I think we can then override this method in the child class and add these error code checks after calling super.checkForFailureStatusCode()
Also I'd do the call to super in the default branch of the the check in child class because superclass always throws an exception, defaulting in undefined error, while child class check can default to calling super method.

jonathan-buttner replied:
Good points. I did some testing and openai will return a 404 and 400 if the model doesn't exist and if the input is malformed. How about for the 422 let's add a TODO and a note saying that openai doesn't return this at the time of writing but we're doing it for mistral which adheres to most of openai's format. The todo can state that we should revisit to decouple them in the future.

Jan-Kazlouski-elastic committed:
Refactor Mistral error response handling and extract StreamingErrorResponse entity

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralUnifiedChatCompletionResponseHandler.java:
```diff
@@ -0,0 +1,51 @@
+ * Handles streaming chat completion responses and error parsing for Mistral inference endpoints.
+ * Adapts the OpenAI handler to support Mistral's simpler error schema with fields like "message" and "http_status_code".
+ */
+public class MistralUnifiedChatCompletionResponseHandler extends OpenAiUnifiedChatCompletionResponseHandler {
```
Done.

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/response/MistralErrorResponseEntity.java:
```diff
@@ -0,0 +1,125 @@
+            XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON)
+                .createParser(XContentParserConfiguration.EMPTY, response.body())
+        ) {
+            var responseMap = jsonParser.map();
```
Done. Also renamed it to MistralErrorResponse because it isn't really an entity. Do let me know if you prefer MistralErrorResponseEntity.

Jan-Kazlouski-elastic commented:
Hi @jonathan-buttner 
I addressed the comments. All fixes are done. I replied to the one comment related to the OpenAI handlers hierarchy https://github.com/elastic/elasticsearch/pull/128538#discussion_r2114309442 to discuss it a bit further.

Also I updated the section related to testing. Three backticks didn't work for me when I was creating initial PR comment, but when I applied them afterwards - worked perfectly. Not sure why, but I will remember for the future PR creation.

Jan-Kazlouski-elastic commented:
New error format:

<details><summary>Create Completion Endpoint</summary>

Not Found:
```
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
        }
    },
    "status": 400
}
```
Unauthorized:
```
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Received an authentication error status code for request from inference entity id [mistral-completion] status [401]. Error message: [{\n  \"message\":\"Unauthorized\",\n  \"request_id\":\"a580d263fb1521778782b22104efb415\"\n}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Received an authentication error status code for request from inference entity id [mistral-completion] status [401]. Error message: [{\n  \"message\":\"Unauthorized\",\n  \"request_id\":\"a580d263fb1521778782b22104efb415\"\n}]"
        }
    },
    "status": 400
}
```
Invalid Model:
```
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Received a bad request status code for request from inference entity id [mistral-completion] status [400]. Error message: [{\"object\":\"error\",\"message\":\"Invalid model: wrong-model-name\",\"type\":\"invalid_model\",\"param\":null,\"code\":\"1500\"}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "status_exception",
            "reason": "Received a bad request status code for request from inference entity id [mistral-completion] status [400]. Error message: [{\"object\":\"error\",\"message\":\"Invalid model: wrong-model-name\",\"type\":\"invalid_model\",\"param\":null,\"code\":\"1500\"}]"
        }
    },
    "status": 400
}
```

</details>

<details><summary>Perform Non-Streaming Completion</summary>

Not Found:
```
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
            }
        ],
        "type": "status_exception",
        "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
    },
    "status": 404
}
```

</details>

<details><summary>Perform Streaming Completion</summary>

Not Found:
```
event: error
data: {"error":{"root_cause":[{"type":"status_exception","reason":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"}],"type":"status_exception","reason":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"},"status":404}
```

</details>

<details><summary>Create Chat Completion Endpoint</summary>

Not Found:
```
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]"
        }
    },
    "status": 400
}
```
Unauthorized:
```
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Received an authentication error status code for request from inference entity id [mistral-chat-completion] status [401]. Error message: [{\n  \"message\":\"Unauthorized\",\n  \"request_id\":\"409ddf538d3f1a55bfe4b7324fe01676\"\n}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Received an authentication error status code for request from inference entity id [mistral-chat-completion] status [401]. Error message: [{\n  \"message\":\"Unauthorized\",\n  \"request_id\":\"409ddf538d3f1a55bfe4b7324fe01676\"\n}]"
        }
    },
    "status": 400
}
```
Invalid Model:
```
{
    "error": {
        "root_cause": [
            {
                "type": "unified_chat_completion_exception",
                "reason": "Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [{\"object\":\"error\",\"message\":\"Invalid model: invalid-model-name\",\"type\":\"invalid_model\",\"param\":null,\"code\":\"1500\"}]"
            }
        ],
        "type": "status_exception",
        "reason": "Could not complete inference endpoint creation as validation call to service threw an exception.",
        "caused_by": {
            "type": "unified_chat_completion_exception",
            "reason": "Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [{\"object\":\"error\",\"message\":\"Invalid model: invalid-model-name\",\"type\":\"invalid_model\",\"param\":null,\"code\":\"1500\"}]"
        }
    },
    "status": 400
}
```

</details>

<details><summary>Perform Streaming Chat Completion</summary>

Not Found:
```
event: error
data: {"error":{"code":"not_found","message":"Resource not found at [https://api.mistral.ai/v1/chat/completions] for request from inference entity id [mistral-chat-completion] status [404]. Error message: [{\"detail\":\"Not Found\"}]","type":"mistral_error"}}
```
Negative Max Tokens:
```
event: error
data: {"error":{"code":"unprocessable_entity","message":"Received an input validation error response for request from inference entity id [mistral-chat-completion] status [422]. Error message: [Input should be greater than or equal to 0]","type":"mistral_error"}}
```
Invalid Model:
```
event: error
data: {"error":{"code":"bad_request","message":"Received a bad request status code for request from inference entity id [mistral-chat-completion] status [400]. Error message: [{\"object\":\"error\",\"message\":\"Invalid model: invalid-model-name\",\"type\":\"invalid_model\",\"param\":null,\"code\":\"1500\"}]","type":"mistral_error"}}
```

</details>

Jan-Kazlouski-elastic committed:
Add Mistral chat completion request and response tests

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralUnifiedChatCompletionResponseHandler.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    protected Exception buildError(String message, Request request, HttpResult result, ErrorResponse errorResponse) {
```
I believe this block is nearly identical for Hugging Face, Mistral, and OpenAI. Could you take a shot at refactoring it to remove the duplication? Maybe we can lift the `instanceof` check out and somehow pass in the `MISTRAL_ERROR` type field as it seems like those are the unique parts.

Could we do that in a separate PR?

Jan-Kazlouski-elastic replied:
Thank you. Fixed.

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/streaming/StreamingErrorResponse.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.Objects;
+import java.util.Optional;
+
+public class StreamingErrorResponse extends ErrorResponse {
```
Can you add a comment with an example error message that this would parse? Let's also add a TODO to note that `ErrorMessageResponseEntity` https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/ErrorMessageResponseEntity.java is nearly identical (doesn't parse as many fields) and we should remove the duplication

Jan-Kazlouski-elastic replied:
Thank you. Done.

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/request/completion/MistralChatCompletionRequestEntity.java:
```diff
@@ -0,0 +1,48 @@
+
+    public MistralChatCompletionRequestEntity(UnifiedChatInput unifiedChatInput, MistralChatCompletionModel model) {
+        this.unifiedChatInput = Objects.requireNonNull(unifiedChatInput);
+        this.unifiedRequestEntity = new MistralUnifiedChatCompletionRequestEntity(unifiedChatInput);
```
Can we adjust this class to follow how we're doing it for openai now: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/request/OpenAiUnifiedChatCompletionRequestEntity.java

Jan-Kazlouski-elastic replied:
Done.

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/response/MistralErrorResponse.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.nio.charset.StandardCharsets;
+
+/**
```
Can you add a few examples of what the error format looks like, or maybe just add them as tests. If you go the test route, just add a comment saying look at the tests for the expected error formats that we're aware of.

Jan-Kazlouski-elastic replied:
Did the change. I also removed the unifiedChatInput for Mistral. While for OpenAI it was left unused along with other fields:
```

    private static final String MODEL_FIELD = "model";
    private static final String MAX_COMPLETION_TOKENS_FIELD = "max_completion_tokens";

    private final UnifiedChatInput unifiedChatInput;
```
    
 I know that refactoring is heavily discouraged by CONTRIBUTING.md, but perhaps I could clean it up for OpenAI as well? Seems like a pretty easy fix.

jonathan-buttner reviewed (changes requested):
Thanks for the changes, left a few suggestions.

Jan-Kazlouski-elastic committed:
Enhance error response documentation and clarify StreamingErrorResponse structure

Jan-Kazlouski-elastic committed:
Refactor Mistral chat completion request handling and introduce skip stream options parameter

Jan-Kazlouski-elastic committed:
Refactor MistralChatCompletionServiceSettings to include rateLimitSettings in equality checks

Jan-Kazlouski-elastic committed:
Enhance MistralErrorResponse documentation with detailed error examples

Jan-Kazlouski-elastic committed:
Add comment for Mistral-specific 422 validation error in OpenAiResponseHandler

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/completion/MistralChatCompletionServiceSettings.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(modelId);
```
Missed that. Thank you. Fixed.

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'origin/main' into feature/mistral-chat-completion-integration

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersions.java

Commit:
[CI] Auto commit changes from spotless

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'origin/main' into feature/mistral-chat-completion-integration

jonathan-buttner reviewed (commented):
Looks good! Left a few suggestions. I think this PR is in a good spot, just ping me when you have the other PR for the refactoring finished so we can merge that one first.

Jan-Kazlouski-elastic committed:
Refactor OpenAiUnifiedChatCompletionRequestEntity to remove unused fields and streamline constructor

Jan-Kazlouski-elastic committed:
Refactor UnifiedChatCompletionRequestEntity and UnifiedCompletionRequest to rename and update stream options parameter

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/unified/UnifiedChatCompletionRequestEntity.java:
```diff
@@ -46,20 +48,13 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-        if (stream) {
-            fillStreamOptionsFields(builder);
+        // If request is streamed and skip stream options parameter is not true, include stream options in the request.
+        if (stream == true && params.paramAsBoolean(SKIP_STREAM_OPTIONS_PARAM, false) == false) {
```
Good thinking. Defaulting boolean to true allows us not to fill it out for every other Took another look at the CONTRIBUTING.md. According to it we should use == check for boolean values only in case we're checking for "false". So I replaced it with:
`stream && params.paramAsBoolean(INCLUDE_STREAM_OPTIONS_PARAM, true)`
Also extended the javadoc for INCLUDE_STREAM_OPTIONS_PARAM

Jan-Kazlouski-elastic committed:
Refactor MistralChatCompletionRequestEntityTests to improve JSON assertion and remove unused imports

Jan-Kazlouski-elastic commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/mistral/MistralUnifiedChatCompletionResponseHandler.java:
```diff
@@ -0,0 +1,51 @@
+/*
```
- I’ve tried this several times and consistently received `read ECONNRESET` errors, these are client-side errors coming from Postman, not server-side ones. By shutting off the client’s internet, we’re preventing it from receiving any further stream data, including both the remaining valid chunks and any potential error payload.
If Mistral does construct an error response, it would still need to reach the client, but in this case, the client is no longer able to receive anything. Unlike Hugging Face, where you might see a structured error midstream.
All errors returned by Mistral so far are strictly non-streaming. The only scenario I can think of where Mistral might emit an error midstream would be due to a genuine server-side malfunction on their part. Or a rate limit, but testing it would be too expensive and even then, the error would probably be non-streaming.

Maybe we could contact Mistral to clarify that?

jonathan-buttner replied:
> Maybe we could contact Mistral to clarify that?

Yeah let's bring this up at the next meeting and see if we can get Serena to follow up.

Jan-Kazlouski-elastic committed:
Add unit tests for MistralUnifiedChatCompletionResponseHandler to validate error handling

Jan-Kazlouski-elastic committed:
Add unit tests for MistralService

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'origin/main' into feature/mistral-chat-completion-integration

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersions.java

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'origin/main' into feature/mistral-chat-completion-integration

# Conflicts:
#	server/src/main/java/org/elasticsearch/TransportVersions.java

jonathan-buttner commented:
Looking good, looks like CI is failing with:

```

REPRODUCE WITH: ./gradlew ":x-pack:plugin:inference:qa:inference-service-tests:javaRestTest" --tests "org.elasticsearch.xpack.inference.InferenceGetServicesIT.testGetServicesWithCompletionTaskType" -Dtests.seed=FB54236D6191C4C3 -Dtests.locale=dz-Tibt-BT -Dtests.timezone=America/Nassau -Druntime.java=24
--
  |  
  | InferenceGetServicesIT > testGetServicesWithCompletionTaskType FAILED
  | java.lang.AssertionError:
  | Expected: <13>
  | but: was <14>
  | at __randomizedtesting.SeedInfo.seed([FB54236D6191C4C3:78524A5E8D7A2437]:0)
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
  | at org.elasticsearch.test.ESTestCase.assertThat(ESTestCase.java:2653)
  | at org.elasticsearch.xpack.inference.InferenceGetServicesIT.testGetServicesWithCompletionTaskType(InferenceGetServicesIT.java:137)


```

jonathan-buttner reviewed (approved):
Thanks for the changes! I did some testing and it looks good.

Jan-Kazlouski-elastic committed:
Update expected service count in testGetServicesWithCompletionTaskType

Jan-Kazlouski-elastic committed:
Merge remote-tracking branch 'origin/main' into feature/mistral-chat-completion-integration

Jan-Kazlouski-elastic commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/128947">](https://github.com/elastic/elasticsearch/pull/128947)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.0 -->
